### PR TITLE
[mlir][spirv] Implement gpu::TargetAttrInterface

### PR DESCRIPTION
**Pull request by silee2**
Create SPIRV Target Attribute to enable GPU compilation pipeline. The Target Attribute is modeled after the existing spriv.target_env Plan is to use this new attribute to enable GPU compilation pipeline for OpenCL kernels only.
The changes do not impact Vulkan shaders using milr-vulkan-runner. New GPU Dialect transform pass spirv-attach-target is implemented for attaching attribute from CLI.
gpu-module-to-binary pass now works with GPU module that has SPIRV module with OpenCL kernel functions inside.

### CODE REVIEW

**Comment by silee2** [comment]
Part of refactoring and breaking down #65539 into smaller PRs.

**Comment by silee2** [comment]
@fabianmcg Let me know if I'm missing anything. I have one more PR after this one that  makes target LLVMIR aware of SPIRV dialect since the new target attribute is part of SPIRV dialect which was not considered a leaf dialect that needs translation to LLVMIR.

**Comment by fabianmcg** [comment]
@silee2 I'll take a closer look later, at first glance it looks like everything is there.

Just one quick question, can you describe the next PR a bit more? Because, you shouldn't need to make the target LLVMIR aware, as  the binary embedding is mostly dialect agnostic, so there shouldn't be an issue with `SPIRVTargetAttr` not having a translation to LLVM.
From what I remember from the original PR the only other modification needed is changing how the kernel is launched, and that change must be done in [Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp ](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp)

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (fabianmcg) [inline mlir/lib/Target/SPIRV/Target.cpp:82]: A check is necessary before `*spvMods.begin()`, as `spvMods` could be an empty range.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (fabianmcg) [inline mlir/lib/Target/SPIRV/Target.cpp:92]: I'd suggest instead of the loop:
```suggestion
  SmallVector<char, 0> spvData(spvBinary.size() * sizeof(uint32_t), 0);
  memcpy(spvData.data(), spvBinary.data(), spvData.size());
```

**Review by fabianmcg — CHANGES_REQUESTED** [review]

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVAttributes.td
Comment (silee2) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVAttributes.td:177]: Those parameters are for describing target environment: https://mlir.llvm.org/docs/Dialects/SPIR-V/#target-environment
The parameters are not used today since lowering to SPIR-V is restricted to common set of SPIR-V ops supported in all environments. 
But in the future, lowering passes can check the parameters and utilize SPIR-V extension ops or do target specific optimizations.

**Comment by silee2** [comment]
> @silee2 I'll take a closer look later, at first glance it looks like everything is there.
> 
> Just one quick question, can you describe the next PR a bit more? Because you shouldn't need to make the target LLVMIR aware, as the binary embedding is mostly dialect agnostic, so there shouldn't be an issue with `SPIRVTargetAttr` not having a translation to LLVM. From what I remember from the original PR the only other modification needed is changing how the kernel is launched, and that change must be done in [Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp ](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp)

**Comment by silee2** [comment]
> @silee2 I'll take a closer look later, at first glance it looks like everything is there.
> 
> Just one quick question, can you describe the next PR a bit more? Because you shouldn't need to make the target LLVMIR aware, as the binary embedding is mostly dialect agnostic, so there shouldn't be an issue with `SPIRVTargetAttr` not having a translation to LLVM. From what I remember from the original PR the only other modification needed is changing how the kernel is launched, and that change must be done in [Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp ](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp)

Yes. SelectObjectAttr.cpp is the only function change.
But the new target attribute "spirv.target" is part of SPIRV dialect.
And SPIRV dialect becomes a new dialect as part of input to LLVMIR translation.
The additional change updates
mlir/include/mlir/Target/LLVMIR/Dialect/All.h
which now includes
registerSPIRVDialectTranslation(registry);
There are no actual SPIRV ops to translate and just a single attribute type.
So the new files
mlir/include/mlir/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.h
and 
mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp
are mostly empty.
Without those changes, mlir-cpu-runner will see spirv.target attribute in the input which is part of SPIRV dialect. And since SPIRV dialect is not registered, mlir-cpu-runner throws an error.
NVVM and ROCDL did not have this issue since NVVM and ROCDL was already registered as an input dialect for LLVMIR translation.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Changed as suggested.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp:87]: Added check for empty range.

**Comment by silee2** [comment]
@fabianmcg Can you take a look again? Answered question and addressed change requests.

**Comment by fabianmcg** [comment]
This patch LGTM! I'll approve once the builds have completed.

> Without those changes, mlir-cpu-runner will see spirv.target attribute in the input which is part of SPIRV dialect. And since SPIRV dialect is not registered, mlir-cpu-runner throws an error.

Did you try just registering the dialect instead of adding a translation? I'm still unsure if adding a translation to LLVM is needed, as the error is probably only happening during parsing? But I guess we can have this conversation on the next PR.

**Comment by silee2** [comment]
check-flang failed on Windows with two unresolved tests. Does not look related to this PR. Merging main again to run CI again.

**Review by fabianmcg — APPROVED** [review]

**Comment by silee2** [comment]
@joker-eph This PR got approval from @fabianmcg. What do I need to do next to get the PR merged?

**Comment by antiagainst** [comment]
Thanks for the contribution and sorry about the delay! I'm getting around to review this pull request (already done a few others). Please wait a bit.

File: mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp
Comment (antiagainst) [inline mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp:57]: We should avoid calling `symbolizeVersion` twice. We can save the result in a local variable in the above call and use it here.

File: mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp
Comment (antiagainst) [inline mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp:77]: Instead doing such explicit loops, just do `walk`?

File: mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp
Comment (antiagainst) [inline mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp:73]: Again, here, we should avoid calling those `symbolize*` functions twice. Just save the result earlier and use it here.

File: mlir/test/Dialect/GPU/spirv-attach-targets.mlir
Comment (antiagainst) [inline mlir/test/Dialect/GPU/spirv-attach-targets.mlir:5]: Can we add tests to show like

1) multiple GPU modules in the same container module?
2) different match regex with hits and misses?

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td
Comment (antiagainst) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:202]: Given the following file

```
Please don't call the symbolize* functions twice everywhere--those functions are expensive. We can save the result in a locla variable and use it later.
```
With `mlir-opt ...`, it will generate:

```
    gpu.module @nvvm_module_1 {...}
    gpu.module @spirv_module_1 [#spirv.target<vce = #spirv.vce<v1.0, [Kernel], []>, resource_limits = <>>] {...}
```

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td
Comment (antiagainst) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:213]: SPIR-V version

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td
Comment (antiagainst) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:215]: s/required/supported/

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:1]: Nit: again, please use SPIR-V in docs and comments. Please make sure everywhere follows the convention. :)

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:41]: SPIR-V implementation of the `gpu:TargetAttrInterface`.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:74]: We don't need this assert here?

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:78]: module->emitError("expected to be a gpu.module op")

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:81]: We could move this upward and check whether it's null instead of the `isa` check at L77.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:84]: Nit: drop `{..}` for trivial if statements per LLVM coding sytle.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:91]: Serialize the spv.module op to SPIR-V blob.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:93]: Nit: "failed to .."

no capitalization to compose better with previous / following messages.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (antiagainst) [inline mlir/lib/Target/SPIRV/Target.cpp:83]: The comment here does not provide additional value; we should remove it or expand it to be more clear.

**Review by antiagainst — CHANGES_REQUESTED** [review]
Thanks again for the contribution! A couple of comments inlined. My main concern is the duplication of `spirv.target` attribute. We should avoid the duplication and confusion there.

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td
Comment (silee2) [reply mlir/include/mlir/Dialect/GPU/Transforms/Passes.td]: Done.

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVAttributes.td
Comment (silee2) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVAttributes.td]: Updated the PR to use spirv::TargetEnvAttr.

File: mlir/include/mlir/Target/SPIRV/Target.h
Comment (silee2) [reply mlir/include/mlir/Target/SPIRV/Target.h]: Done.

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVAttributes.td
Comment (silee2) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVAttributes.td]: Done.

File: mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp
Comment (silee2) [reply mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp]: Done.

File: mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp
Comment (silee2) [reply mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp]: Done.

File: mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp
Comment (silee2) [reply mlir/lib/Dialect/GPU/Transforms/SPIRVAttachTarget.cpp]: Done.

File: mlir/test/Dialect/GPU/spirv-attach-targets.mlir
Comment (silee2) [reply mlir/test/Dialect/GPU/spirv-attach-targets.mlir]: Done.

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td
Comment (silee2) [reply mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:203]: Done.

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td
Comment (silee2) [reply mlir/include/mlir/Dialect/GPU/Transforms/Passes.td]: Done.

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td
Comment (silee2) [reply mlir/include/mlir/Dialect/GPU/Transforms/Passes.td]: Done.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Done.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Done.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Removed.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Done.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Done.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Done.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Done.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Done. Thanks for the tip.

File: mlir/lib/Target/SPIRV/Target.cpp
Comment (silee2) [reply mlir/lib/Target/SPIRV/Target.cpp]: Removed.

**Comment by silee2** [comment]
@antiagainst Thanks for the detailed review and helpful comments! Updated the PR accordingly.

**Comment by silee2** [comment]
@antiagainst I'm waiting for your approval.

**Review by antiagainst — APPROVED** [review]
LGTM. Thanks for addressing the comments!

**Comment by antiagainst** [comment]
I revised the commit message a bit and landed it. Thanks for the waiting!

**Comment by JOE1994** [comment]
This revision seems related to the following buildbot failure:

* https://lab.llvm.org/buildbot/#/builders/268/builds/1747
  (+ builds 1748 & 1749) 

Would you mind taking a look? Thank you 👍

**Comment by antiagainst** [comment]
Sorry about that! I uploaded https://github.com/llvm/llvm-project/commit/4a4b8570f7c16346094c59fab1bd8debf9b177e1 to fix this. Verified that it fixes the break for me locally.

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.h
Comment (jpienaar) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.h:18]: This feels off: you have GPU dialect transforms depending on SPIRV attributes/dialect (esp in header). Why is this pass in GPU dialect rather than SPIRV one?

Comment (fabianmcg) [reply mlir/include/mlir/Dialect/GPU/Transforms/Passes.h:18]: Back when the compilation redesign was happening we decided to add the attach* passes in GPU to avoid polluting lower level dialects with GPU includes. However, I do agree that include shouldn't be there, as far as I could tell, that include is only needed by one pass option `mlir::spirv::TargetEnvAttr::kUnknownDeviceID`, so it should be possible to remove it. I'll fix it.

Comment (antiagainst) [reply mlir/include/mlir/Dialect/GPU/Transforms/Passes.h:18]: SG. Thanks @jpienaar & @fabianmcg!

**Pull request merged**
